File: claudesync/sessions.py
"""Read Claude Code chat session transcripts (.jsonl files) so a backup or
an imported one can be previewed and verified before you actually restore
it or start trusting it.

Claude Code's transcript format isn't a documented, stable API, so this
parser is deliberately tolerant: it extracts a role and readable text where
it can recognize the shape, and quietly skips whatever it doesn't
recognize (a malformed line, an unfamiliar block type) rather than
crashing. Worst case you see less detail, not an error.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Iterator

SNIPPET_LENGTH = 140


@dataclass
class SessionInfo:
    path: Path
    project_hint: str
    size: int
    mtime: float
    message_count: int
    snippet: str


def _extract_text(content) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for block in content:
            if not isinstance(block, dict):
                continue
            block_type = block.get("type")
            if block_type == "text" and isinstance(block.get("text"), str):
                parts.append(block["text"])
            elif block_type == "tool_use":
                parts.append(f"[used tool: {block.get('name', 'tool')}]")
            elif block_type == "tool_result":
                parts.append("[tool result]")
        return "\n".join(p for p in parts if p)
    return ""


def iterate_messages(jsonl_path: Path) -> Iterator[dict]:
    """Yield {"role", "text", "timestamp"} dicts, skipping unreadable lines."""
    try:
        lines = Path(jsonl_path).read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError:
        return

    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue
        if not isinstance(obj, dict):
            continue

        message = obj.get("message")
        role = None
        text = ""
        if isinstance(message, dict):
            role = message.get("role")
            text = _extract_text(message.get("content"))
        if not role:
            role = obj.get("type")
        if not role and not text:
            continue

        yield {"role": role or "unknown", "text": text.strip(), "timestamp": obj.get("timestamp")}


def _snippet(jsonl_path: Path) -> tuple[str, int]:
    count = 0
    snippet = ""
    for msg in iterate_messages(jsonl_path):
        count += 1
        if not snippet and msg["text"]:
            flat = " ".join(msg["text"].split())
            snippet = flat[:SNIPPET_LENGTH] + ("…" if len(flat) > SNIPPET_LENGTH else "")
    return snippet, count


def list_sessions(root: Path) -> list[SessionInfo]:
    """Find every .jsonl session transcript under `root` (a backup, an
    imported directory, or a live ~/.claude directory), newest first.
    """
    root = Path(root).expanduser()
    if not root.is_dir():
        return []

    infos = []
    for path in sorted(root.rglob("*.jsonl")):
        try:
            stat = path.stat()
        except OSError:
            continue
        snippet, count = _snippet(path)
        infos.append(
            SessionInfo(
                path=path,
                project_hint=path.parent.name,
                size=stat.st_size,
                mtime=stat.st_mtime,
                message_count=count,
                snippet=snippet,
            )
        )
    infos.sort(key=lambda i: i.mtime, reverse=True)
    return infos


def render_text(jsonl_path: Path, limit: int | None = None) -> Iterator[str]:
    """Yield formatted "[role] text" lines for a transcript, for printing."""
    count = 0
    for msg in iterate_messages(jsonl_path):
        line = f"[{msg['role']}] {msg['text']}" if msg["text"] else f"[{msg['role']}]"
        yield line
        count += 1
        if limit and count >= limit:
            yield f"... (truncated at {limit} messages; use --limit to see more)"
            break


def render_markdown(jsonl_path: Path, title: str) -> str:
    """Render a whole transcript as a readable Markdown document."""
    lines = [f"# Chat session: {title}", ""]
    any_messages = False
    for msg in iterate_messages(jsonl_path):
        any_messages = True
        role = (msg["role"] or "unknown").replace("_", " ").capitalize()
        lines.append(f"**{role}**")
        lines.append("")
        lines.append(msg["text"] if msg["text"] else "*(no text)*")
        lines.append("")
    if not any_messages:
        lines.append("*(no readable messages found in this transcript)*")
    return "\n".join(lines)


_SLUG_RE = re.compile(r"[^A-Za-z0-9._-]+")


def _slugify(text: str) -> str:
    slug = _SLUG_RE.sub("-", text).strip("-")
    return slug or "session"


def export_all_sessions(root: Path, out_dir: Path, fmt: str = "md") -> list[Path]:
    """Write every chat session found under `root` as one readable file per
    session in `out_dir` (markdown by default, or plain text), so the whole
    chat history can be browsed or archived without ClaudeSync or the raw
    .jsonl format.
    """
    if fmt not in ("md", "txt"):
        raise ValueError(f"Unknown export format: {fmt!r} (expected 'md' or 'txt')")

    out_dir = Path(out_dir).expanduser()
    out_dir.mkdir(parents=True, exist_ok=True)

    written: list[Path] = []
    used_names: set[str] = set()
    for info in list_sessions(root):
        date = datetime.fromtimestamp(info.mtime).strftime("%Y-%m-%d")
        base = _slugify(f"{date}_{info.project_hint}_{info.path.stem}")
        name = f"{base}.{fmt}"
        suffix = 2
        while name in used_names:
            name = f"{base}-{suffix}.{fmt}"
            suffix += 1
        used_names.add(name)

        if fmt == "md":
            content = render_markdown(info.path, title=info.path.stem)
        else:
            content = "\n".join(render_text(info.path)) or "(no readable messages found in this transcript)"

        out_path = out_dir / name
        out_path.write_text(content, encoding="utf-8")
        written.append(out_path)

    return written

File: desktop/src/core/sessions.js
'use strict';

/**
 * Read Claude Code chat session transcripts (.jsonl files) so a backup or
 * an imported one can be previewed and verified before you actually
 * restore it or start trusting it.
 *
 * Claude Code's transcript format isn't a documented, stable API, so this
 * parser is deliberately tolerant: it extracts a role and readable text
 * where it can recognize the shape, and quietly skips whatever it doesn't
 * recognize (a malformed line, an unfamiliar block type) rather than
 * throwing. Worst case you see less detail, not a crash.
 */

const fs = require('fs');
const path = require('path');

const SNIPPET_LENGTH = 140;

function extractText(content) {
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) {
    const parts = [];
    for (const block of content) {
      if (!block || typeof block !== 'object') continue;
      if (block.type === 'text' && typeof block.text === 'string') {
        parts.push(block.text);
      } else if (block.type === 'tool_use') {
        parts.push(`[used tool: ${block.name || 'tool'}]`);
      } else if (block.type === 'tool_result') {
        parts.push('[tool result]');
      }
    }
    return parts.filter(Boolean).join('\n');
  }
  return '';
}

/** Parses transcript text into {role, text, timestamp} entries, skipping unreadable lines. */
function parseMessages(content) {
  const messages = [];
  for (const rawLine of content.split('\n')) {
    const line = rawLine.trim();
    if (!line) continue;

    let obj;
    try {
      obj = JSON.parse(line);
    } catch {
      continue;
    }
    if (!obj || typeof obj !== 'object' || Array.isArray(obj)) continue;

    let role = null;
    let text = '';
    if (obj.message && typeof obj.message === 'object') {
      role = obj.message.role || null;
      text = extractText(obj.message.content);
    }
    if (!role) role = obj.type || null;
    if (!role && !text) continue;

    messages.push({ role: role || 'unknown', text: text.trim(), timestamp: obj.timestamp ?? null });
  }
  return messages;
}

function walkJsonlFiles(root) {
  const results = [];
  function walk(dir) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) walk(full);
      else if (entry.isFile() && entry.name.endsWith('.jsonl')) results.push(full);
    }
  }
  walk(root);
  return results.sort();
}

function snippetFor(messages) {
  for (const msg of messages) {
    if (msg.text) {
      const flat = msg.text.replace(/\s+/g, ' ').trim();
      return flat.length > SNIPPET_LENGTH ? `${flat.slice(0, SNIPPET_LENGTH)}…` : flat;
    }
  }
  return '';
}

/** Finds every .jsonl session transcript under `root`, newest first. */
function listSessions(root) {
  if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) return [];

  const infos = walkJsonlFiles(root).map((filePath) => {
    const stat = fs.statSync(filePath);
    const messages = parseMessages(fs.readFileSync(filePath, 'utf8'));
    return {
      path: filePath,
      projectHint: path.basename(path.dirname(filePath)),
      size: stat.size,
      mtimeMs: stat.mtimeMs,
      messageCount: messages.length,
      snippet: snippetFor(messages),
    };
  });

  infos.sort((a, b) => b.mtimeMs - a.mtimeMs);
  return infos;
}

/** Reads a transcript's messages, capped at `limit` (default: all). */
function readSession(filePath, limit) {
  const messages = parseMessages(fs.readFileSync(filePath, 'utf8'));
  if (!limit || messages.length <= limit) {
    return { messages, total: messages.length, truncated: false };
  }
  return { messages: messages.slice(0, limit), total: messages.length, truncated: true };
}

function renderMarkdown(messages, title) {
  const lines = [`# Chat session: ${title}`, ''];
  for (const msg of messages) {
    const role = (msg.role || 'unknown').replace(/_/g, ' ');
    const capitalized = role.charAt(0).toUpperCase() + role.slice(1);
    lines.push(`**${capitalized}**`, '', msg.text || '*(no text)*', '');
  }
  if (messages.length === 0) lines.push('*(no readable messages found in this transcript)*');
  return lines.join('\n');
}

function renderPlainText(messages) {
  if (messages.length === 0) return '(no readable messages found in this transcript)';
  return messages.map((m) => (m.text ? `[${m.role}] ${m.text}` : `[${m.role}]`)).join('\n');
}

function slugify(text) {
  const slug = text.replace(/[^A-Za-z0-9._-]+/g, '-').replace(/^-+|-+$/g, '');
  return slug || 'session';
}

/**
 * Writes every chat session found under `root` as one readable file per
 * session in `outDir` (markdown by default, or plain text), so the whole
 * chat history can be browsed or archived without ClaudeSync or the raw
 * .jsonl format.
 */
function exportAllSessions(root, outDir, format = 'md') {
  if (format !== 'md' && format !== 'txt') {
    throw new Error(`Unknown export format: ${format} (expected 'md' or 'txt')`);
  }

  fs.mkdirSync(outDir, { recursive: true });

  const written = [];
  const usedNames = new Set();
  for (const info of listSessions(root)) {
    const date = new Date(info.mtimeMs).toISOString().slice(0, 10);
    const stem = path.basename(info.path, path.extname(info.path));
    const base = slugify(`${date}_${info.projectHint}_${stem}`);

    let name = `${base}.${format}`;
    let suffix = 2;
    while (usedNames.has(name)) {
      name = `${base}-${suffix}.${format}`;
      suffix += 1;
    }
    usedNames.add(name);

    const messages = parseMessages(fs.readFileSync(info.path, 'utf8'));
    const content = format === 'md' ? renderMarkdown(messages, stem) : renderPlainText(messages);

    const outPath = path.join(outDir, name);
    fs.writeFileSync(outPath, content, 'utf8');
    written.push(outPath);
  }

  return written;
}

module.exports = { parseMessages, listSessions, readSession, exportAllSessions };
